mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +347,206 @@ propagateUnmergeVectorization(T &&dimAndLength,
   return result;
 }
 
+// Key data structures for vectorization analysis
+using DimToMergeMap =
+    llvm::DenseMap<uint32_t,
+                   std::tuple<TransformMapAttr, TransformAttr, size_t>>;
+
+using ContiguousMergesMap =
+    llvm::DenseMap<std::pair<TransformMapAttr, TransformAttr>,
+                   SmallVector<SmallVector<uint32_t>>>;
+
+void findCountiguousGroupsUnmerge(const ArrayRef<uint32_t> upperDims,
+                                  const ArrayRef<int64_t> params,
+                                  DimToMergeMap &dimToMerge,
+                                  ContiguousMergesMap &contiguousGroups) {
+  size_t i = 0;
+  while (i < upperDims.size()) {
+    // The i-th dimension (i.e., dimension upperDims[i])
+    // uses params[i]
+
+    auto keyI = dimToMerge[upperDims[i]];
+    auto mergePairI = std::make_pair(std::get<0>(keyI), std::get<1>(keyI));
+    if (!mergePairI.second) {
+      i++;
+      continue;
+    }
+
+    SmallVector<uint32_t> groupCandidate;
+    SmallVector<size_t> dimPosition;
+    for (size_t j = i; j < upperDims.size(); j++) {
+
+      // Unit lengths don't affect the mergeability logic
+      if (params[j] == 1) {
+        continue;
+      }
+
+      auto keyJ = dimToMerge[upperDims[j]];
+      auto mergePairJ = std::make_pair(std::get<0>(keyJ), std::get<1>(keyJ));
+
+      if (!mergePairJ.second) {
+        break;
+      }
+      auto mergePairJParams = mergePairJ.second.getParams();
+      auto mergePairJDims = mergePairJ.second.getLowerDims();
+      size_t posJ = std::get<2>(keyJ);
```

Comment:
Since we're dealing with these rather dense tuples ... would they make sense as structs instead?